support/k8s/spire-adm-webhook/webhook.go

```diff
@@ -0,0 +1,225 @@
+	if w.c.PodLabel != "" {
+		// the webhook has been been configured with a pod label. if the pod
+		// has that label, use the value to construct the pod entry. otherwise
+		// ignore the pod altogether.
+		if labelValue, ok := pod.Labels[w.c.PodLabel]; ok {
+			return w.createPodEntryByLabel(ctx, pod, w.c.PodLabel, labelValue)
+		}
+		return nil
+	}
+
+	// the webhook has not been configured with a pod label. create an entry
+	// based on the service account.
+	return w.createPodEntryByServiceAccount(ctx, pod)
+}
+
+func (w *Webhook) createPodEntryByLabel(ctx context.Context, pod *corev1.Pod, labelKey, labelValue string) error {
+	return w.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: w.nodeID(),
+		SpiffeId: w.makeID("%s", labelValue),
+		Selectors: []*common.Selector{
+			namespaceSelector(pod.Namespace),
+			podNameSelector(pod.Name),
+		},
+	})
+}
+
+func (w *Webhook) createPodEntryByServiceAccount(ctx context.Context, pod *corev1.Pod) error {
+	return w.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: w.nodeID(),
+		SpiffeId: w.makeID("ns/%s/sa/%s", pod.Namespace, pod.Spec.ServiceAccountName),
+		Selectors: []*common.Selector{
+			namespaceSelector(pod.Namespace),
+			podNameSelector(pod.Name),
+		},
+	})
+}
+
+func (w *Webhook) deletePodEntry(ctx context.Context, namespace, name string) error {
+	log := w.c.Log.WithFields(logrus.Fields{
+		"ns":  namespace,
+		"pod": name,
+	})
+
+	entries, err := w.c.R.ListBySelectors(ctx, &common.Selectors{
+		Entries: []*common.Selector{
+			namespaceSelector(namespace),
+			podNameSelector(name),
+		},
+	})
+	if err != nil {
+		return errs.New("unable to list by pod entries: %v", err)
+	}
+
+	log.WithField("count", len(entries.Entries)).Debug("Deleting entries")
+
+	var errGroup errs.Group
+	for _, entry := range entries.Entries {
+		_, err := w.c.R.DeleteEntry(ctx, &registration.RegistrationEntryID{
+			Id: entry.EntryId,
+		})
+		if err != nil {
+			log.WithField("err", err.Error()).Debug("Failed deleting pod entry")
+			errGroup.Add(errs.New("unable to delete entry %q: %v", entry.EntryId, err))
+		}
+	}
+	return errGroup.Err()
+}
+
+func (w *Webhook) nodeID() string {
+	return w.makeID("node")
+}
+
+func (w *Webhook) makeID(pathFmt string, pathArgs ...interface{}) string {
+	id := url.URL{
+		Scheme: "spiffe",
+		Host:   w.c.TrustDomain,
+		Path:   path.Clean(fmt.Sprintf(pathFmt, pathArgs...)),
+	}
+	return id.String()
+}
+
+func (w *Webhook) createEntry(ctx context.Context, entry *common.RegistrationEntry) error {
+	// ensure there is a node registration entry for PSAT nodes in the cluster.
+	log := w.c.Log.WithFields(logrus.Fields{
+		"parent_id": entry.ParentId,
+		"spiffe_id": entry.SpiffeId,
+		"selectors": selectorsField(entry.Selectors),
+	})
+	_, err := w.c.R.CreateEntry(ctx, entry)
+	switch status.Code(err) {
+	case codes.OK, codes.AlreadyExists:
+		log.Debug("Created entry")
+		return nil
+	default:
+		log.WithField("err", err).Debug("Failed to create entry")
```

Comment:
This should probably be logged at error level

Reply:
👍